### PR TITLE
refactor!: Use inheritance for type-erasure of Magnetic Field Providers (1/2)

### PR DESCRIPTION
**Pull request by paulgessinger**
This PR creates a new interface called `MagneticFieldProvider`. It contains virtual methods `getField` and `getFieldGradient`, with and without a cache argument.
The idea is to use this instead of concrete B field provider types. To retain the opaque cache type functionality, caches now use a type-erased common type,
so they can be expressed in the interface. 
The interface also has a virtual method `makeCache`, which returns an instance of the concrete cache type, wrapped in the opaque type.

I added two possible implemenations of the cache type: one based on `std::any`, and one which
allocates a small chunk of memory on the stack and uses that. The opaque cache
type has an interface that allows consistent usage of either, but a typedef in
`BFieldProvider` chooses what's used. I benchmarked them against each other
using the generic propagation example and didn't see significant performance
changes between them, and relative to master:

| test | any | sbo | master|
|:--|:--|:--|:--|
| prop | 38.09 ms/event |37.74 ms/event  |37.11 ms/event |
| prop cov | 75.09 ms/event |74.29 ms/event | 74.44 ms/event |

(this seems to be within measurement precision)

I also benchmarked the interpolated bfield test, and the solenoid field bench, which didn't show any differences.
I did **NOT** switch out the concrete types in the steppers (yet), because I'd like feedback first. So right now, the 
steppers still know the concrete type, and the compiler might be able to do some advanced optimization of the virtual calls.
If we want to continue with this, I'll strip the steppers of the concrete type and make them use the field through a type-erased interface instance, and 
check back if that degrades performance (I don't expect it to).

Ultimately: I want to leverage this to simplify the propagator and B field handling in the examples.

BREAKING CHANGE: This changes the interface that all magnetic field providers need to comply with. It now uses a virtual base class.

### CODE REVIEW

**Comment by paulgessinger** [comment]
@asalzburger, @msmk0 

Maybe @HadrienG2  has some thoughts as well?

**Comment by HadrienG2** [comment]
Seems reasonable to me. We shouldn't template things over the entire world if we don't have a measurable performance motivation to do so, otherwise things like the CKF happen.

I'd need a more in-depth study to figure out why the runtime performance difference is so small (assuming it stays small in your future tests), but my basic assumption is that the code that is being measured here is not bottlenecked by any of the new overheads that were introduced (memory allocation, virtual function calls, pointer indirection...), or that the compiler manages to optimize out some of the overheads (e.g. as currently written, with all implementations in headers, the code should lend itself relatively well to devirtualization).

**Comment by paulgessinger** [comment]
I'm having trouble getting the `std::any` based variant to compile on GCC9, maybe related to [this bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90415). I'd propose just using my custom opaque type. Right now it has a fixed size of 512 bytes, and will refuse to compile if you plug in something larger. We can extend the size, or switch to *true* SBO (as in heap allocates if size is too large) in the future if necessary.

Thoughts?

**Comment by msmk0** [comment]
Minor detail: our naming is a bit inconsistent; the module is named `MagneticField` and most providers use some variant of `BField`. Since I find `MagneticField` to be a bit clearer, could the new interface just be named `MagneticField` as well?

**Comment by msmk0** [comment]
I also like the idea of the small object cache and would conceptually prefer it over `std::any` since we have full control over it and we avoid allocations in most cases (and it seems to be a bit faster). If you decide to use that, my suggestion would be to place in into `Utilities/detail` and rename without the abbreviation, e.g. `SmallObjectCache` or `SmallOpaqueObjectCache`.

**Comment by paulgessinger** [comment]
Hmm. I see your point about the name, but considering the deployment / connector use case for example in ATLAS, I believe *provider* is the better name overall. I could live with `MagneticFieldProvider`, but that's another mouth full.

**Comment by paulgessinger** [comment]
Concerning the SBO, right now, `SBOCache` is an implementation detail, and is typedef'd in `BFieldProvider`. The interface hides the way it's actually stored, so I don't think its strictly necessary to generalize this just yet. When we have another use case for it, it should be straight forward to generalize.

**Comment by msmk0** [comment]
> I misread what you wrote. I can move it to another header if you want, but I thought it makes sense to keep it close.

I was assuming that the implementation in the header was the first prototype and that you would retain only the cache implementation version that we end up using in the end. I'll let you decide whether to keep it close or move it.

> Hmm. I see your point about the name, but considering the deployment / connector use case for example in ATLAS, I believe provider is the better name overall. I could live with MagneticFieldProvider, but that's another mouth full.

The `Provider` naming sounds ATLAS-specific. Conceptually, it is just an interface. E.g. the constant field is just a field, there is not connector/provider functionality needed. `IMagneticField` is another option, but I really do not like this for reasons outlined e.g. [here](https://developer.okta.com/blog/2019/06/25/iinterface-considered-harmful).

**Comment by paulgessinger** [comment]
I have moved/renamed `SBOCache` to `SmallObjectCache` in a separate header in a detail folder.

Concerning the `BFieldProvider`, having thought about it more, I think *provider* is actually the better word, since it abstracts whether the instance **is** the field or just provides access to a field.

If anyone else wants to weigh in here, they're welcome to.

**Comment by paulgessinger** [comment]
With the all-clear by @baschlag in #675, this is ready for review.

**Comment by msmk0** [comment]
> Concerning the `BFieldProvider`, having thought about it more, I think _provider_ is actually the better word, since it abstracts whether the instance **is** the field or just provides access to a field. This applies to both our internal implementations, as well as connector ones, like we deployed in Athena.

@paulgessinger I can understand that reasoning. So `Provider` it shall be. Would you consider going with `MagneticFieldProvider` to match with the module name? I know it is a bit longer, but it would be more consistent.

**Comment by paulgessinger** [comment]
Done.

File: Core/include/Acts/MagneticField/InterpolatedBFieldMap.hpp [hunk 1]
```diff
@@ -268,82 +269,73 @@ class InterpolatedBFieldMap final {
...
   }
 
-  /// @brief retrieve magnetic field value & its gradient
-  ///
-  /// @param [in]  position   global 3D position
-  /// @param [out] derivative gradient of magnetic field vector as (3x3) matrix
-  /// @param [in,out] cache Cache object. Contains field cell used for
-  /// @return magnetic field vector
+  /// @copydoc BFieldBase::getFieldGradient(const
+  /// Vector3&,ActsMatrix<3,3>&,BFieldBase::Cache&)
```
Comment (paulgessinger) [inline Core/include/Acts/MagneticField/InterpolatedBFieldMap.hpp:328]: Does this work despite the linebrack?

Comment (asalzburger) [reply Core/include/Acts/MagneticField/InterpolatedBFieldMap.hpp:328]: Guess not ...

**Review by asalzburger — APPROVED** [review]
I was not sure if we had agreed on moving (at least partially) virtual interfaces to a `IName.hpp` - otherwise looks good to me.

File: Core/include/Acts/MagneticField/MagneticFieldProvider.hpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+#include "Acts/MagneticField/MagneticFieldContext.hpp"
+#include "Acts/MagneticField/detail/SmallObjectCache.hpp"
+
+#include <array>
+#include <memory>
+
+namespace Acts {
+
+/// Base class for all magnetic field providers
+class MagneticFieldProvider {
```
[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/MagneticField/MagneticFieldProvider.hpp:21]: I don't think we do. @msmk0 argued against it, and I don't have a strong opinion.

Your other comment shows one of my docstrings, which I didn't update when renaming. I'll fix those

[the remark this answers is not in the document]
Comment (paulgessinger) [reply Core/include/Acts/MagneticField/MagneticFieldProvider.hpp:21]: I fixed the docstrings.

**Comment by asalzburger** [comment]
Merging master in, and then wee'll merge - after yesterday's discussion.

**Pull request merged**